Add AutonomousActionQueue

JacobLikesBagels opened the pull request:
AutonoumousActionQueue is a class that stores and manages autonomous actions. Once in the queue, they can be executed in order with a variety of methods. 

This might all be useless and a massive waste of time, but I am not sure. I just think having a system to make multiple auton routines other than just calling a whole bunch of methods seems like a good idea. I'm not sure

JacobLikesBagels committed:
Add AutonomousActionQueue

AutonoumousActionQueue is a class the stores and manages autonomous actions. Once in the queue, they can be executed in order with a variety of methods. it might be helpful? or maybe a waste of time, but I don't really know

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+     */
+    public void executeNext() {
+
+        if (queuePosition >= queue.size() - 1)
+            return;
```
Colin's biggest rule

always, always, always, ALWAYS put brackets

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+     * @param amountOfTimes The amount of times to execute the current action
+     */
+    public void executeCurrentActionNTimes(int amountOfTimes) {
+        for (int i = 0; i < amountOfTimes; i++)
+            executeAction(getCurrentAction());
```
ABB (Always be bracketing)

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    private Action getCurrentAction() {
+        if (queuePosition >= queue.size())
+            queuePosition = queue.size() - 1;
```
ABB (Always be bracketing)

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+     * @param millisecondDelay the delay between each action in milliseconds
+     */
+    public void executeAll(int millisecondDelay) {
+        if (millisecondDelay < 0)
+            millisecondDelay = 0;
```
ABB (Always be bracketing)

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+     * @param millisecondDelay the delay between each action in milliseconds
+     */
+    public void executeRemaining(int millisecondDelay) {
+        if (millisecondDelay < 0)
+            millisecondDelay = 0;
```
ABB (Always be bracketing)

Nanosplitter commented on src/main/java/frc/robot/AutonomousActionQueue.java:
```diff
@@ -0,0 +1,213 @@
+     * @param position the position in queue
+     */
+    public void setQueuePosition(int position) {
+        if (position > queue.size() || position < 0)
+            return;
```
ABB (Always be bracketing)

Nanosplitter commented on src/test/java/AutonomousActionQueueTest.java:
```diff
@@ -0,0 +1,78 @@
+public class AutonomousActionQueueTest {
+
+    @Nested
+    class Iforgor {
```
Uh

Nanosplitter commented on src/test/java/AutonomousActionQueueTest.java:
```diff
@@ -0,0 +1,78 @@
+
+        // @Test
+        // There isn't a good way to make sure something waits long enough, so this test sucks
+        public void executeAllDelayTest() {
+            queue.queue(ActionType.DO_NOTHING, 0, 0);
+            queue.queue(ActionType.DO_NOTHING, 0, 1);
+            queue.queue(ActionType.DO_NOTHING, 0, 2);
+            queue.executeAll(1000);
+
+        }
```
Not sure what this method is doing

JacobLikesBagels replied:
Yeah I forgot to remove that part.

Nanosplitter reviewed (changes requested):
Y'all are gonna give me a heart attack with your hatred of brackets

JacobLikesBagels committed:
Add Brackets + Wait action

Add some brackets so I always be bracketing :sunglasses:  Also added WAIT as one of the available actions.

JacobLikesBagels committed:
Remove useless test

Remove that silly little test that does nothing but take up room.

nstringham commented:
I think this is overall a good idea but I have a few thoughts on how it could be improved.

It seams to me like we might want to do something in the future that doesn't make sense to describe using only a speed and an amount. Additionally it seams like as this system becomes more complex it might get messy to have all the logic in a single switch statement. One possible improvement could be to abstract and instruction into a Interface or Base Class and then have methods and data on the Class and construct an action and then pass it into the queue function.

I noticed you are using an ArrayList as a queue. You might be interested in using one of the queues built into Java. In particular [ArrayDeque](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html)

IMO it would make your code more readable if you always included `this.` when using an instance variable.

I don't know what `Timer.delay` is but it seams like it may be blocking which means we can't do anything else. It would likely be better to use something that doesn't block. In the past we have done this by making a function on the AutonomousActionQueue class that needs to be called inside of autonomousPeriodic.

It might be interesting to look at what I did in 2020.
https://github.com/WTMCRobotics/2020-InfiniteRecharge/blob/5b6b7605f8b6815ebaa3c008af6c6b9611b863e2/src/main/java/frc/robot/Robot.java#L517-L531
Definitely don't copy everything from 2020 though your solution is better in a lot of ways.
